src/DynamicProjectOptions.cmake

```diff
@@ -0,0 +1,146 @@
+    "ENABLE_CPPCHECK\;OFF\;${MAKEFILE_OR_NINJA}\;Enable cppcheck analysis during compilation"
+    "ENABLE_IPO\;OFF\;OFF\;Enable whole-program optimization"
+    "ENABLE_INCLUDE_WHAT_YOU_USE\;OFF\;OFF\;Enable include-what-you-use analysis during compilation"
+    "ENABLE_PCH\;ON\;OFF\;Enable pre-compiled-headers support"
+    "ENABLE_DOXYGEN\;OFF\;OFF\;Build documentation with Doxygen"
+    "ENABLE_USER_LINKER\;OFF\;OFF\;Allow custom linker settings"
+    "ENABLE_BUILD_WITH_TIME_TRACE\;OFF\;OFF\;Generates report of where compile-time is spent"
+    "ENABLE_UNITY\;OFF\;OFF\;Merge C++ files into larger C++ files, can speed up compilation sometimes"
+    "ENABLE_SANITIZER_LEAK\;OFF\;OFF\;Make memory leaks into hard runtime errors"
+    "ENABLE_SANITIZER_THREAD\;OFF\;OFF\;Make thread race conditions into hard runtime errors"
+    "ENABLE_SANITIZER_MEMORY\;OFF\;OFF\;Make other memory errors into runtime errors")
+
+foreach(option ${options})
+  list(
+    GET
+    option
+    0
+    option_name)
+  list(
+    GET
+    option
+    1
+    option_user_default)
+  list(
+    GET
+    option
+    2
+    option_developer_default)
+  list(
+    GET
+    option
+    3
+    option_description)
+
+  if(DEFINED ${option_name}_DEFAULT)
+    if(DEFINED ${option_name}_DEVELOPER_DEFAULT OR DEFINED ${option_name}_USER_DEFAULT)
+      message(
+        SEND_ERROR
+          "You have separately defined user/developer defaults and general defaults for ${option_name}. Please either provide a general default OR separate developer/user overrides"
+      )
+    endif()
+
+    set(option_user_default ${${option_name}_DEFAULT})
+    set(option_developer_default ${${option_name}_DEFAULT})
+  endif()
+
+  if(DEFINED ${option_name}_USER_DEFAULT)
+    set(option_user_default ${${option_name}_USER_DEFAULT})
+  endif()
+
+  if(DEFINED ${option_name}_DEVELOPER_DEFAULT)
+    set(option_developer_default ${${option_name}_DEVELOPER_DEFAULT})
+  endif()
+
+  cmake_dependent_option(
+    OPT_${option_name}
+    "${option_description}"
+    ${option_developer_default}
+    ENABLE_DEVELOPER_MODE
+    ${option_user_default})
+
+  if(OPT_${option_name})
+    set(${option_name}_VALUE ${option_name})
+  else()
+    unset(${option_name}_VALUE)
+  endif()
+endforeach()
+
+macro(dynamic_project_options)
+  project_options(
+    ${ENABLE_CONAN_VALUE}
+    ${ENABLE_CACHE_VALUE}
+    ${WARNINGS_AS_ERRORS_VALUE}
+    ${ENABLE_CPPCHECK_VALUE}
+    ${ENABLE_CLANG_TIDY_VALUE}
+    ${ENABLE_COVERAGE_VALUE}
+    ${ENABLE_IPO_VALUE}
+    ${ENABLE_INCLUDE_WHAT_YOU_USE_VALUE}
+    ${ENABLE_PCH_VALUE}
+    ${ENABLE_DOXYGEN_VALUE}
+    ${ENABLE_USER_LINKER_VALUE}
+    ${ENABLE_BUILD_WITH_TIME_TRACE_VALUE}
+    ${ENABLE_UNITY_VALUE}
+    ${ENABLE_SANITIZER_ADDRESS_VALUE}
+    ${ENABLE_SANITIZER_LEAK_VALUE}
+    ${ENABLE_SANITIZER_UNDEFINED_BEHAVIOR_VALUE}
+    ${ENABLE_SANITIZER_THREAD_VALUE}
+    ${ENABLE_SANITIZER_MEMORY_VALUE}
+    ${ARGN})
```

Comment:
Does this ARGN allow writing the options in the CMake file like how it is done for `project_options`? If that's the case, it is worth mentioning in the readme example.

Reply:
It's there for the `PCH_HEADERS` options and `*_WARNING` overrides. If other overrides of options work, it is by accident.